### PR TITLE
schema: add OVS link aggregation

### PR DESCRIPTION
**Pull request by phoracek**
Signed-off-by: Petr Horacek <phoracek@redhat.com>

### CODE REVIEW

**Comment by phoracek** [comment]
@EdDev @cathay4t @tyll could you please take a look?

File: libnmstate/schemas/operational-state.yaml [hunk 1]
```diff
@@ -287,6 +287,23 @@ definitions:
                     type: string
                   access-tag:
                     type: string
+                link-aggregation:
```
Comment (EdDev) [inline libnmstate/schemas/operational-state.yaml]: I do not understand the indentation. Is it supposed to be at the same level as the previous `access-tag` key?

But the main question at this level is: How will we differentiate between a regular port and a link-aggregation one? Please consider that nmstate likes partial editing in general.

File: libnmstate/schemas/operational-state.yaml [hunk 2]
```diff
@@ -287,6 +287,23 @@ definitions:
...
+                  type: object
+                  required:
+                  - mode
+                  - slaves
```
Comment (EdDev) [inline libnmstate/schemas/operational-state.yaml]: If both are required, we cannot do partial editation (now or in the future).
I suggest you leave it up to the implementation to validate this, per its ability.

File: libnmstate/schemas/operational-state.yaml [hunk 3]
```diff
@@ -287,6 +287,23 @@ definitions:
...
+                  properties:
+                    mode:
+                      type: string
+                      enum:
+                      - balance-slb
```
Comment (EdDev) [inline libnmstate/schemas/operational-state.yaml]: Do we want to control this at the schema level? It will make it hard to expend the list in the future.
Adding an interface type is one thing, but limiting a link aggregation mode of an ovs port is something else.

File: libnmstate/schemas/operational-state.yaml [hunk 4]
```diff
@@ -287,6 +287,23 @@ definitions:
...
+                    slaves:
+                      type: array
+                      minItems: 2
+                      items:
+                        type: object
+                        properties:
```
Comment (EdDev) [inline libnmstate/schemas/operational-state.yaml]: Can you please give an example of an additional property? If you can add one now, it will make this much clearer why this layer is needed (instead of a list of iface names).

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: link-aggretation port has to have two or more interfaces attached. I don't see how that breaks the partial editation, could you elaborate?

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: There is a **lot** of OVS interface attributes available http://www.openvswitch.org/support/dist-docs/ovs-vswitchd.conf.db.5.html. However, we may just set them in a separate interface entry. We can change it to list of strings to match the linux bonding (although it is scary to limit ourselves).

File: libnmstate/schemas/operational-state.yaml (hunk 2 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: Ack.

File: libnmstate/schemas/operational-state.yaml (hunk 3 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: I don't see how that is an issue, but I'm fine making it more free.

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: You use the existence of the subtree do determine if the user requested a bond or not. This will not work well with partial editation. If the user  wants to change a port vlan for example, he will not mention the subtree.. but then, you will not be able to differentiate that request from just making that port a non bond. I think we need here an ‘enable’ flag.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: That won't work the other way around though. In OVS database, the only mark that something is a bond is that multiple interfaces are under the same port. Nothing else, no attributes, nothing.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: Ah, now I get you. So after all we should have kept the `type` of port :)

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: You may misunderstood me.
I was looking for an argument to leave it like this. It will be hard to change the api from a list of strings to list of objects in the future.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: I don't have any additional attribute to add now.

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: Any suggestion how it should look like? @EdDev

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: I actually remember what I thought at the time of that review.
I thought that we can define the ports as they now and can set link aggregation as a group of other ports. That will work with keeping the slaves as strings.
Makes sense?

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: We can still check if it is a bond or simple port while applying a partial config. No need for an additional attribute. While generating proxy interfaces, we can check from the system for the current number of ifaces. That is what I did in https://github.com/nmstate/nmstate/pull/514 if I remember correctly.

Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: I have an idea: we can declare that for changing a bond to a non bond, it needs to be “absent” first. That will work.
But we now need an absent property.

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: I am strongly against that (as I raised with you before). The main issue is that the `port` entry we have now has attributes of an OVS `port`, e.g. VLAN.

If I understood you correctly, you want to have `link-aggregation` type linking multiple of these `ports`. That would mean `port` alone is an OVS `port` and `interface`, while `port` under `link-aggregation` is an OVS `interface` only, although it has all attributes of a port. It would make the API way more complicated and hard to read what is defined in a desired state.

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: > We can still check if it is a bond or simple port while applying a partial config. No need for an additional attribute. While generating proxy interfaces, we can check from the system for the current number of ifaces. That is what I did in #514 if I remember correctly.

The problem is not about what exists currently, but what the client intentions are.
But I think I answered it earlier now.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: I don't get what you are trying to solve. Changing bond to a non-bond can look like following:

```yaml
port:
- name: bond1
  link-aggregation:
    slaves:
      - name: eth0
      - name: eth1
```

To:

```yaml
port:
- name: eth0
```

You cannot partially change list of slaves. There is no such support in linux bonding nor linux bridge afaik.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: @EdDev please give me a specific example of a change that you think would not work with the current schema and I will tell you how would it be tackled or why it should not be supported.

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: I think you got it right. The suggestion is merge port anf interface settings.. which fits all cases except bonding. For bonding, the enslaved ports will ignore the port settings. I think this is a better model.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: You want to setup interface details? Do it in respective interface `ethernet` or `ovs-internal`. OVS `bridge` `port` list should have `port` specific attributes only. There is this shorthand of "if you defined only name, we implicitly add an interface with the same name" which does not fit the OVS DB schema.

What do you think is a better model again?

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: Your example should cause a failure and not change a state. If in practice tries to define a contradicting state: on one part eth0 is a port, in parallel it is a bond slave.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: Ah, you want a partial editing within the port? Oof...

Any idea how to tackle it (other than creating link-aggregation type and making nmstate into a circus)?

File: libnmstate/schemas/operational-state.yaml (hunk 4 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: I think ovs has interface properties that apply to eth0 and alike as well. Those are relevant for ovs, not for eth0.
I will try to fetch an example later today.

Comment (phoracek) [reply libnmstate/schemas/operational-state.yaml]: Please also try to write an example of the API proposed by you.

File: libnmstate/schemas/operational-state.yaml (hunk 1 above)
Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: > I have an idea: we can declare that for changing a bond to a non bond, it needs to be “absent” first. That will work.
> But we now need an absent property.

Comment (tyll) [reply libnmstate/schemas/operational-state.yaml]: > > I have an idea: we can declare that for changing a bond to a non bond, it needs to be “absent” first. That will work.
> > But we now need an absent property.

IMHO this violates the idea that Nmstate users declare what they want instead of how they get it. Also, where do you want to declare this?

Comment (EdDev) [reply libnmstate/schemas/operational-state.yaml]: This is what we do with routes and rules, so I do not know what this violates.
I will open a discussion on the PR root on this with an example.

**Comment by EdDev** [comment]
Please see below my proposal.

Basic concepts that guided me:
- Nmstate does not want to use two layers of ports and interfaces.
  Ports and interfaces are therefore merged into a single layer: ports.
  For the special case of bonds, a virtual port is defined with a link-aggregation subtree.
- As ports have "weight" on them with potentially complex config, adding and removing them individually would be preferable (like be do with routes and rules).
  The implementation may start with full list-only, but I am not sure it will be simple to implement, just as we saw with the routes case.
- Nmstate allows partial editation whenever possible, we should try and keep this working.
  When trying to hint about some property through an existence of another, this editation may break.

Define a new bond port (showing only the port subtree):
```yaml
port:
  - name: eth0
    lacp-port-id: 12345 # OVS Interface parameter, given as an example
  - name: eth1
    lacp-port-id: 6789 # OVS Interface parameter, given as an example
  - name: bond1
    link-aggregation:
      slaves:
        - eth0
        - eth1 
```
Consider the bond as a virtual port aggregator that groups ports.

Port name lookup algorithm:
- Is it a kernel iface?
- Is it an internal interface?
- Is it a bond port? i.e. defined on the bridge already, or has a link-aggregation subtree (for new ones).

To detach a port from the bridge, including the removal of a bond with its slaves:
```yaml
port:
  - name: bond1
    state: absent
```
This will also remove the slaves implicitly.
We could delay this implementation later. Defining the full-state first and only later the partial is possible, although I find it less convenient.

**Comment by phoracek** [comment]
@EdDev could you please add a northbound port to the example?

**Comment by phoracek** [comment]
@EdDev VLAN in OVS is an attribute of a `port`. What if `eth0` is set with `vlan 10` and `eth1` with `vlan 20`, how are we supposed to merge that into a single bond port?

**Comment by EdDev** [comment]
@phoracek , I added those details.
I'll keep editing it per your requests to make it simple to follow.

**Comment by phoracek** [comment]
@EdDev since the PR I send is just a schema and the schema you proposed is completely different, do you mind taking over this PR creating a new one? I see no benefit in rewriting your comments into a yaml schema. I would review your PR.

I really don't like the idea of explaining someone when you can use which attributes on port and when not and testing the whole corner-case of link-aggregation handling. But whatever API you guys consider nmstate-ish the most, I just want my bond.

**Comment by EdDev** [comment]
> I really don't like the idea of explaining someone when you can use which attributes on port and when not and testing the whole corner-case of link-aggregation handling.

Raising concerns is exactly what the schema design is all about. Explaining users and treating edge cases exists in all options.
The idea is to be specific and point us to the pros/cons of each option.
You are more than welcome to provide the alternative with examples so we can discuss why it is better or not.
As an example, I want to understand what is the alternative to add an interface property that may fit both a slave-port and a regular port.

> But whatever API you guys consider nmstate-ish the most, I just want my bond.

I prefer well over fast, even if well is slow.
Please address the examples and goals to help us make a good long lasting API.

**Comment by phoracek** [comment]
By "someone"  meant users via documentation.

My main concern is that we have a list what looks like a flat list of ports (the current state, simple to understand). What you propose is to add a hierarchy to those ports. I found that really unintuitive. Both our algorithm and users need to crawl to that list and build the image of actual topology themselves (instead of it being obvious from yaml).

I still want to brainstorm a little on my proposal, where you have a list of `ports`. They can be either simple (without aggregation), meaning they represent an `interface` of the same name. Or they can have explicitly set `link-aggregation`, meaning they cover multiple `interfaces`. Can we add `absent` just to that?

Back to my original example in https://github.com/nmstate/nmstate/pull/562#discussion_r351871736. With `absent`, it would be:

```yaml
port:
- name: bond1
  link-aggregation:
    slaves:
      - name: eth0
      - name: eth1
```

Changed into:

```yaml
port:
- name: eth0
- name: bond1
  state: absent
```

**Comment by EdDev** [comment]
Your suggestion is ok, but I am missing one thing:
> As an example, I want to understand what is the alternative to add an interface property that may fit both a slave-port and a regular port.

**Comment by phoracek** [comment]
@EdDev I'd say that would be specific and open to discussion for every single attribute. I don't think we can apply a  generic rule for that. The only rule I'd start with is that you cannot have an interface mentioned in two ports (name and slave).

For your two examples, VLAN would be always only on the top level. LACP port id would be under `link-aggregation`.

Some attributes may need to be described in API on both places, but when `link-aggregation` is in use, they are allowed only on the slave level.

**Comment by EdDev** [comment]
@phoracek please use the lacp example I gave and show it.
We have a bunch of properties already, lets sample some and see how it looks.

In this regard, IMO you are proposing a simple model for the simple case and a complex inconsistent model for the special cases.
I have a strange feeling that you are taking us back to model this the same way ovs does it.. port having interfaces.. at least in their model they do not have duplication.

As the only point left is this interface property, can we delay the decision on it without the need to change api later?

**Comment by phoracek** [comment]
@EdDev the issue is that the current API is hiding interface implicitly in the port. That was OK and simple for simple cases, but is not feasible for more complex cases, such as bonding ;)

```yaml
port:
- name: bond1
  vlan: ... #  OVS generic port property
  link-aggregation:
    mode: balance-slb  # OVS bonding port property
    lacp-system-id: 1000  # OVS bonding port property
    slaves:
      - name: eth0
        lacp-port-id: 1  # OVS bonding interface property
      - name: eth1
        lacp-port-id: 2  # OVS bonding interface property
```

Is there an interface property that would make sense for both standalone port and bonding, while not being configured directly in `ethernet` or `internal-ovs` types?

**Comment by EdDev** [comment]
Ok, lets try this approach for now. I will dig into the interface properties later today and check if something may break.

Please post the schema and all relevant parts as done with the sriov patch: https://github.com/nmstate/nmstate/pull/592/files

**Comment by phoracek** [comment]
@EdDev should be done. I won't be able to address your comments until tomorrow evening. Thanks.

File: libnmstate/schema.py [hunk 5]
```diff
@@ -244,6 +244,16 @@ class OVSBridge(object):
...
 
     PORT_SUBTREE = 'port'
-    PORT_NAME = 'name'
-    PORT_VLAN_MODE = 'vlan-mode'
-    PORT_ACCESS_TAG = 'access-tag'
+
+    class Port(object):
+        NAME = 'name'
+        VLAN_MODE = 'vlan-mode'
+        ACCESS_TAG = 'access-tag'
```
Comment (EdDev) [inline libnmstate/schema.py]: This is not a related change and causes a collisions with other PR/s.
Please focus on the added new properties:
```
class Port:
    LINK_AGGREGATION_SUBTREE = 'link-aggregation'
    ...
```

File: libnmstate/schema.py [hunk 6]
```diff
@@ -244,6 +244,16 @@ class OVSBridge(object):
...
+        LINK_AGGREGATION_SUBTREE = 'link-aggregation'
+
+        class LinkAggregation(object):
```
Comment (EdDev) [inline libnmstate/schema.py]: Please drop the `(object)` for new classes.

File: tests/integration/ovs_test.py [hunk 7]
```diff
@@ -103,3 +104,16 @@ def test_create_and_remove_ovs_bridge_with_internal_port_and_static_ip():
...
     assertlib.assert_absent(BRIDGE1)
     assertlib.assert_absent(PORT1)
+
+
+@pytest.mark.xfail(
+    raises=NmstateNotImplementedError,
+    reason='OVS link-aggregation is not supported yet',
+    strict=True,
+)
+def test_la_not_implemented():
```
Comment (EdDev) [inline tests/integration/ovs_test.py]: Please use `linkaggregation` instead of `la`.

But it does not make sense to test it, and in fact you expect an xfail, which means you are supposed to test the creation of a link-aggregation. Please show that and present how it is supposed to be used.

File: tests/integration/testlib/ovslib.py [hunk 8]
```diff
@@ -47,6 +47,13 @@ def set_options(self, options):
     def add_system_port(self, name):
         self._add_port(name)
 
+    # TODO: Implement adding slaves and mode
+    def add_la_port(self, name):
```
Comment (EdDev) [inline tests/integration/testlib/ovslib.py]: How about `add_linkaggregation_port`.

And if you define this helper, it makes sense to use all defined properties in the schema (i.e. the mode and slaves). Otherwise, you can just drop this.

File: tests/lib/schema_validation_test.py [hunk 9]
```diff
@@ -408,3 +409,28 @@ def test_non_interger_route_table(self, default_data):
         with pytest.raises(js.ValidationError) as err:
             libnmstate.validator.validate(default_data)
         assert 'main' in err.value.args[0]
+
+
+class TestOVSBridge(object):
```
Comment (EdDev) [inline tests/lib/schema_validation_test.py]: Drop the `(object)` please.

File: tests/lib/schema_validation_test.py [hunk 10]
```diff
@@ -26,11 +26,12 @@
 from libnmstate.schema import Constants
 from libnmstate.schema import DNS
 from libnmstate.schema import Ethernet
-from libnmstate.schema import LinuxBridge as LB
-from libnmstate.schema import VXLAN
 from libnmstate.schema import Interface
 from libnmstate.schema import InterfaceType
+from libnmstate.schema import LinuxBridge as LB
```
Comment (EdDev) [inline tests/lib/schema_validation_test.py]: This one and VXLAN do not seem to be related to this change.

File: libnmstate/nm/applier.py [hunk 11]
```diff
@@ -390,6 +391,11 @@ def _build_connection_profile(iface_desired_state, base_con_profile=None):
             )
             settings.append(linux_bridge_setting)
     elif iface_type == ovs.BRIDGE_TYPE:
+        for port in iface_desired_state.get(OvsB.CONFIG_SUBTREE, {}).get(
+            OvsB.PORT_SUBTREE, []
+        ):
+            if OvsB.Port.LINK_AGGREGATION_SUBTREE in port:
+                raise NmstateNotImplementedError
```
Comment (EdDev) [inline libnmstate/nm/applier.py]: This is not the place we raise the missing implementation.
Missing implementations are usually raised in the setting creation.
For example: https://github.com/nmstate/nmstate/blob/95af61535301ee885d415f5853aa2c06c23972ad/libnmstate/nm/bridge.py#L93

File: tests/integration/testlib/ovslib.py [hunk 12]
```diff
@@ -58,9 +68,11 @@ def add_internal_port(self, name, ipv4_state):
         )
 
     def _add_port(self, name):
+        port = {OVSBridge.PORT_NAME: name}
         self._bridge_iface[OVSBridge.CONFIG_SUBTREE].setdefault(
             OVSBridge.PORT_SUBTREE, []
-        ).append({OVSBridge.PORT_NAME: name})
+        ).append(port)
+        return port
```
Comment (EdDev) [inline tests/integration/testlib/ovslib.py]: it is not recommended to do both setting and getting in the same method.
It may be convenient when used, but it causes other problems.
The method changes the state. The caller can fetch the new state or re-calculate `{OVSBridge.PORT_NAME: name}`.

File: tests/lib/schema_validation_test.py [hunk 13]
```diff
@@ -571,3 +571,28 @@ def test_non_interger_route_table(self, default_data):
         with pytest.raises(js.ValidationError) as err:
             libnmstate.validator.validate(default_data)
         assert 'main' in err.value.args[0]
+
+
+class TestOVSBridge:
```
Comment (EdDev) [inline tests/lib/schema_validation_test.py]: Please be specific, like the vlan one is.
